client/gc_test.go

```diff
@@ -343,136 +341,53 @@ func TestAllocGarbageCollector_MakeRoomForAllocations_GC_Fallback(t *testing.T)
-	}
-
-	// Create a job
-	state := server.State()
-	job := mock.Job()
-	job.TaskGroups[0].Tasks[0].Driver = "mock_driver"
-	job.TaskGroups[0].Tasks[0].Config = map[string]interface{}{
-		"run_for": "30s",
-	}
-	nodeID := client.Node().ID
-	if err := state.UpsertJob(98, job); err != nil {
-		t.Fatalf("error upserting job: %v", err)
-	}
-	if err := state.UpsertJobSummary(99, mock.JobSummary(job.ID)); err != nil {
-		t.Fatalf("error upserting job summary: %v", err)
-	}
-
-	newAlloc := func() *structs.Allocation {
-		alloc := mock.Alloc()
-		alloc.JobID = job.ID
-		alloc.Job = job
-		alloc.NodeID = nodeID
-		return alloc
-	}
-
-	// Create the allocations
-	allocs := make([]*structs.Allocation, 7)
-	for i := 0; i < len(allocs); i++ {
-		allocs[i] = newAlloc()
-	}
+	logger := testlog.HCLogger(t)
+	statsCollector := &MockStatsCollector{}
+	conf := gcConfig()
+	conf.MaxAllocs = 3
+	gc := NewAllocGarbageCollector(logger, statsCollector, &MockAllocCounter{}, conf)
 
-	// Upsert a copy of the allocs as modifying the originals later would
-	// cause a race
-	{
-		allocsCopy := make([]*structs.Allocation, len(allocs))
-		for i, a := range allocs {
-			allocsCopy[i] = a.Copy()
-		}
-		if err := state.UpsertAllocs(100, allocsCopy); err != nil {
-			t.Fatalf("error upserting initial allocs: %v", err)
-		}
-	}
+	ar1, cleanup1 := allocrunner.TestAllocRunnerFromAlloc(t, mock.Alloc())
+	defer cleanup1()
+	ar2, cleanup2 := allocrunner.TestAllocRunnerFromAlloc(t, mock.Alloc())
+	defer cleanup2()
+	ar3, cleanup3 := allocrunner.TestAllocRunnerFromAlloc(t, mock.Alloc())
+	defer cleanup3()
 
-	// 7 total, 0 GC'd
-	assertAllocs(7, 0)
+	go ar1.Run()
+	go ar2.Run()
+	go ar3.Run()
 
-	// Set the first few as terminal so they're marked for gc
-	const terminalN = 4
-	for i := 0; i < terminalN; i++ {
-		// Copy the alloc so the pointers aren't shared
-		alloc := allocs[i].Copy()
-		alloc.DesiredStatus = structs.AllocDesiredStatusStop
-		allocs[i] = alloc
-	}
-	if err := state.UpsertAllocs(101, allocs[:terminalN]); err != nil {
-		t.Fatalf("error upserting stopped allocs: %v", err)
-	}
+	exitAllocRunner(ar1)
 
-	// 7 total, 1 GC'd to get down to limit of 6
-	assertAllocs(7, 1)
+	gc.MarkForCollection(ar1.Alloc().ID, ar1)
+	gc.MarkForCollection(ar2.Alloc().ID, ar2)
+	gc.MarkForCollection(ar3.Alloc().ID, ar3)
 
-	// Add one more alloc
-	if err := state.UpsertAllocs(102, []*structs.Allocation{newAlloc()}); err != nil {
-		t.Fatalf("error upserting new alloc: %v", err)
+	{
+		alloc := mock.Alloc()
+		err := gc.MakeRoomFor([]*structs.Allocation{alloc})
+		require.NoError(t, err)
 	}
 
-	// 8 total, 1 GC'd to get down to limit of 6
-	// If this fails it may be due to the gc's Run and MarkRoomFor methods
-	// gc'ing concurrently. May have to disable gc's run loop if this test
-	// is flaky.
-	assertAllocs(8, 2)
+	// We GC a single alloc runner.
+	require.Equal(t, true, ar1.IsDestroyed())
```

Comment:
At the top of the function: `require := require.New(t)` and then use `require.True(ar1.IsDestroyed())` and `require.False(xxx)`

Reply:
@dantoml It is a common pattern across the entire code base and in this case uniformity is a benefit. This can be a larger discussion but in the mean time can you please conform to the projects standards

Reply:
sure 👍